Feature/bc 282 assessed totals

sumachalla39 opened the pull request:
## What is this PR?

[Link to story](https://dsdmoj.atlassian.net/browse/BC-282)

Added one e2e test(AssessedClaimDetailsTest) for reduced-still-escaped outcome, still need to add few more e2e tests after the right data available.

<img width="1318" height="1008" alt="image" src="https://github.com/user-attachments/assets/5adf9350-d466-4def-bd8e-a283de3f6212" />


## Checklist

Before you ask people to review this PR:

- [x] Branch naming followed as per [LAA Ways Of Working](https://dsdmoj.atlassian.net/wiki/spaces/LP1/pages/5697536341/LAA+Ways+of+working#Core-Branches).
- [x] Tests should be passing: `./gradlew test`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [x] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [x] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

jackheslop96 commented on src/main/resources/templates/review-and-amend.html:
```diff
@@ -44,8 +44,9 @@ <h2 class="govuk-summary-card__title" th:text="#{claimSummary.claimCostsTableTit
                 <tbody class="govuk-table__body">
                 <tr class="govuk-table__row"
-                    th:each="row : ${viewModel.tableRows}"
+                    th:each="row : ${viewModel.tableRows}" th:if="${!viewModel.isTotalRow(row)}"
```
Since `getTableRows` already has some logic to display `getTotalAmount` or not, could we just pass a boolean into the `getTableRows` method (e.g. `boolean isClaimSummary`) so we keep the logic to hide the row in one place. Also means we don't need to determine `isTotalRow` on each row in the list. Would also be good to see this unit tested.

sumachalla39 replied:
getTableRows works well for claim-summary,  where it display only when claim has no assessment.  Review And Amend page shouldn't display total as per design. 

I have added PageType, hope this helps for future customisation for page specific

jackheslop96 commented on src/main/resources/templates/claim-summary.html:
```diff
@@ -1,92 +1,166 @@
-            <div class="govuk-width-container">
-                <div class="govuk-grid-row">
-                    <div class="govuk-grid-column-two-thirds">
+    )}"
```
Have there been any functional changes here or is it just indentation changes? Hard to identify where the actual diffs are.

sumachalla39 replied:
Latest Main seems to be reformatted, I will update this

jackheslop96 commented on src/main/java/uk/gov/justice/laa/amend/claim/config/security/SecurityConstants.java:
```diff
@@ -10,6 +10,7 @@ public class SecurityConstants {
         "/webjars/**",
-        "/favicon.ico"
+        "/favicon.ico",
+        "/webjars/*"
```
Out of interest, does `/webjars/**` not already cover this through recursive matching?

sumachalla39 replied:
accidental change, fixed it

sumachalla39 committed:
BC-282 Trying to pass e2e tests with flaky data

sumachalla39 committed:
BC-282 Disabled e2e test

sumachalla39 committed:
BC-282 Added add or UpdateButton click to avoid issues for when assessment exist not exist due to reuse of data

sumachalla39 committed:
BC-282 Modified Assessed Totals visibility

sumachalla39 committed:
BC-282 Modified Assessed Totals visibility

sumachalla39 committed:
BC-282 Modified Assessed Totals visibility

sumachalla39 committed:
BC-282 Assessed Totals, Allowed Totals, e2e tests

# Conflicts:
#	.github/workflows/test.yml
#	src/e2e/main/java/uk/gov/justice/laa/amend/claim/pages/ReviewAndAmendPage.java

jackheslop96 commented on src/main/java/uk/gov/justice/laa/amend/claim/controllers/ChangeAssessedTotalsController.java:
```diff
@@ -40,7 +40,7 @@ public String onPageLoad(
         ClaimDetails claim = (ClaimDetails) request.getAttribute(claimId);
 
-        if (claim.getAssessedTotalVat().getStatus() == ClaimFieldStatus.DO_NOT_DISPLAY || claim.getAssessedTotalInclVat().getStatus() == ClaimFieldStatus.DO_NOT_DISPLAY) {
+        if (claim.getAssessedTotalVat().getStatus() == ClaimFieldStatus.NOT_MODIFIABLE || claim.getAssessedTotalInclVat().getStatus() == ClaimFieldStatus.NOT_MODIFIABLE) {
```
Is `DO_NOT_DISPLAY` now obsolete?

sumachalla39 replied:
We might need to wait and see if it can be used for fields like bolton when it has null to hide them

jackheslop96 commented on src/main/java/uk/gov/justice/laa/amend/claim/utils/ThymeleafUtils.java:
```diff
@@ -90,6 +90,12 @@ public ThymeleafString getAssessedValue(ClaimField value) {
     }
 
+    public ThymeleafString getChangeLinkText(ClaimField value) {
+        var val = value != null && value.getAssessed() == null ? "service.add" : "service.change";
```
Consider creating `isAssessed` method in ClaimField, as we have `assessed == null`, `assessed != null`, etc. in a couple different places.

sumachalla39 replied:
This can be refactored later and used only to display label, is it worth to add as part of ClaimField?

jackheslop96 commented on src/main/java/uk/gov/justice/laa/amend/claim/validators/FeeCodeValidator.java:
```diff
@@ -21,7 +21,7 @@ private FeeCodeValidator() {
     public static boolean isNotValidFeeCode(ClaimDetails claimDetails) {
         if (!(claimDetails instanceof CrimeClaimDetails)) {
-            return false;
+            return true;
```
For readability can we remove the `!` and return false?

jackheslop96 commented on src/main/java/uk/gov/justice/laa/amend/claim/viewmodels/ClaimDetailsView.java:
```diff
@@ -54,28 +54,44 @@ default Map<String, Object> getSummaryRows() {
+    private ClaimField resolveValue(ClaimField assessed, ClaimField allowed) {
+        if (assessed != null && allowed != null && assessed.getStatus() == ClaimFieldStatus.NOT_MODIFIABLE) {
+            assessed.setAssessed(allowed.getAssessed());
+            return assessed;
```
This return can be removed

jackheslop96 commented on src/main/resources/templates/claim-summary.html:
```diff
@@ -48,32 +48,80 @@ <h2 class="govuk-summary-card__title" th:text="#{claimSummary.values}"></h2>
+                                                th:attr="aria-disabled=${disabled}"
                                                 th:disabled="${disabled}"
-                                                th:attr="aria-disabled=${disabled}">
+                                                th:data-testid="claimDetailsAssessmentButton"
```
For consistency, can we use lowercase with hyphens for ids / test ids

jackheslop96 commented on src/main/resources/templates/review-and-amend.html:
```diff
@@ -105,44 +107,35 @@ <h2 class="govuk-summary-card__title" th:text="#{claimSummary.assessedTotalsTabl
+                        <td class="govuk-table__cell govuk-!-font-weight-bold" th:text="#{${row.label}}"></td>
+                        <td class="govuk-table__cell" th:text="${@ThymeleafUtils.getFormattedValue(row.calculated).resolve(#messages)}"></td>
+                        <td class="govuk-table__cell" th:text="${@ThymeleafUtils.getFormattedValue(row.submitted).resolve(#messages)}"></td>
+                        <td class="govuk-table__cell" th:if="${(row.assessed != null)}" th:text="${@ThymeleafUtils.getFormattedValue(row.assessed).resolve(#messages)}"></td>
```
Re. one of the earlier comments, could also make use of an `isAssessed` boolean in ClaimField here

jackheslop96 commented on wiremock/mappings/get-claim-mappings.json:
```diff
@@ -1,7 +1,8 @@
   "request": {
     "method": "GET",
-    "urlPattern": "/api/v0/submissions/(.*)/claims/(.*)"
+    "urlPattern": "/api/v0/submissions/(?!ddc07f4f|3fa85f64)[^/]*/claims/(?!0199a474-65d5-7f95-a1f3-290976326ecd|0199c9a4-64ec-79c6-a292-e250278f5f71).*"
```
Since we only have one claim mapping, is there a need for this change?

sumachalla39 replied:
This was originally intened to add different mocks for each claim type, later on I have moved from mock to use real data reverted back

jackheslop96 commented on wiremock/__files/get-submission-response.json:
```diff
@@ -1,29 +0,0 @@
-{
```
Do we not still need the wiremock for getting a submission? I'd have thought the claim summary age would fail to load?

jackheslop96 commented on wiremock/mappings/get-submission-mappings.json:
```diff
@@ -1,13 +0,0 @@
-{
```
As above

sumachalla39 committed:
BC-282 Review comments

sumachalla39 committed:
BC-282 Review comments

sumachalla39 committed:
BC-282 Review comments

sumachalla39 committed:
BC-282 Changed Assessed Totals to check status

jackheslop96 commented on src/test/java/uk/gov/justice/laa/amend/claim/views/ViewTestBase.java:
```diff
@@ -53,6 +55,10 @@ protected Document renderDocument() throws Exception {
   protected Document renderDocument(Map<String, Object> variables) throws Exception {
+    claim.setAssessmentOutcome(OutcomeType.REDUCED_TO_FIXED_FEE);
+    claim.getAssessedTotalVat().setStatus(ClaimFieldStatus.MODIFIABLE);
+    claim.getAssessedTotalInclVat().setStatus(ClaimFieldStatus.MODIFIABLE);
```
Worth moving these to `setup()` ?
